build/provenance.go

```diff
@@ -0,0 +1,198 @@
+package build
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/containerd/containerd/content/proxy"
+	"github.com/docker/buildx/util/confutil"
+	"github.com/docker/buildx/util/progress"
+	controlapi "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/client"
+	provenancetypes "github.com/moby/buildkit/solver/llbsolver/provenance/types"
+	"github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+func setRecordProvenance(ctx context.Context, c *client.Client, sr *client.SolveResponse, ref string, pw progress.Writer) error {
+	mode := confutil.MetadataProvenance()
+	if mode == confutil.MetadataProvenanceModeNone {
+		return nil
+	}
+	pw = progress.ResetTime(pw)
+	return progress.Wrap("resolve build record provenance", pw.Write, func(l progress.SubLogger) error {
+		bo := backoff.NewExponentialBackOff()
+		bo.MaxElapsedTime = 5 * time.Second
+		return backoff.Retry(func() error {
+			res, err := fetchProvenance(ctx, c, ref, mode, l)
```

Comment:
I'm not sure if the logging of blob pulls is needed here. These are(is) always small file so I think it is more noise than useful info.

Reply:
This is mainly to display provenance size in log output and also per platform for multi-platform builds so we could use it as metrics as follow-up of @jsternberg's work. Would also be useful for Builds view in Docker Desktop (Build timing charts).

Reply:
Still these are small file pulls from the daemon, not from a remote source like registry. There isn't much difference in these values just being part of the solverresponse map or pulled from another blob. Single line about writing metadata file should be enough I think.